File: 2019/01/day1.nim
import strutils, sugar, sequtils, math, strutils


func compute(x: float): int =
    result = int(floor(x / 3.0) - 2)

func compute(x: int): int =
  result = compute(x.float)

func compute(x: string): int =
  if x.len > 0:
    result = compute(parseFloat(x))
  else:
    result = 0

func computeWithFuel(x: string): int =
  var values = newSeq[int]()
  var fuel = compute(x)
  values.add(fuel)
  while fuel >= 0:
    fuel = compute(fuel)
    if fuel > 0:
      values.add(fuel)
  result = sum(values)

when isMainModule:
  let input = readFile("./input.txt").string.splitLines
  echo "part 1"
  echo sum(input.map(x => compute(x)))

  echo "part 2"
  echo sum(input.map(x => computeWithFuel(x)))

File: 2019/01/day1.py
#!/usr/bin/env python
import math


def compute(x: int) -> int:
    return math.floor((x / 3.0)) - 2


def compute_with_fuel(x: int) -> int:
    remainder = compute(x)
    fuels = [remainder]
    while remainder >= 0:
        remainder = compute(remainder)
        if remainder >= 0:
            fuels.append(remainder)

    return sum(fuels)


if __name__ == '__main__':
    with open('input.txt', 'r') as fp:
        raw_modules = fp.readlines()

    modules = [int(x) for x in raw_modules if len(x) > 0]

    print("part 1")
    print(sum([compute(x) for x in modules]))

    print("part 2")
    print(sum([compute_with_fuel(x) for x in modules]))
